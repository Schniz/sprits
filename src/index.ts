export * as Step from "./step";
