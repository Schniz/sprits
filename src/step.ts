import { Context, Deferred, Effect } from "effect";
import { type NodeModel, digraph, toDot as toGraphvizDot } from "ts-graphviz";

type ConstructionError<Inputs extends AnyStep[], R> = [
	{
		[key in keyof Inputs]: Inputs[key]["name"];
	}[Extract<keyof Inputs, number>],
	R,
] extends [infer Provided, StepContext<infer Requested>]
	? Exclude<Requested, Provided> extends never
		? never
		: string extends Requested
			? never
			: `undeclared step input ${Requested}`
	: never;

const ConstructionError = Symbol("ConstructionError");

export function make<
	const Name extends string,
	const Inputs extends AnyStep[],
	A,
	E,
	R,
	CE = ConstructionError<Inputs, R>,
>(
	...[opts]: [CE] extends [never]
		? [Omit<Step<Name, A, E, R | Current, Inputs>, "read">]
		: [
				Omit<Step<Name, A, E, R | Current, Inputs>, "read"> & {
					run: Step<Name, A, E, R | Current, Inputs>["run"] & [CE];
				},
			]
): Step<Name, A, E, R, Inputs> {
	const run = opts.run.pipe(
		Effect.provideService(Current, { title: String(opts.name) }),
	);
	return {
		...opts,
		run,
		read: Context.GenericTag<StepContext<Name>, A>(`step/${opts.name}`),
	} as any;
}

const Info = Symbol("info");

export type Info<S extends AnyStep> = NonNullable<S[typeof Info]>;

type ImmediateProvidedContexts<
	S extends AnyStep,
	Inputs extends AnyStep[] = Info<S>["Inputs"],
> = { [key in keyof Inputs]: StepContext<Inputs[key]["name"]> }[Extract<
	keyof Inputs,
	number
>];

type RequiredContext<
	S extends AnyStep,
	Inputs extends AnyStep[] = [S, ...Info<S>["Inputs"]],
	Current = never,
> = {
	empty: Current;
	nonempty: RequiredContext<
		S,
		[...Tail<Inputs>, ...Inputs[0]["inputs"]],
		| Current
		| Exclude<
				Effect.Effect.Context<Inputs[0]["run"]>,
				ImmediateProvidedContexts<Inputs[0], Inputs[0]["inputs"]>
		  >
	>;
}[Inputs extends [] ? "empty" : "nonempty"];
// Effect.Effect.Context<S["run"]>
// Exclude<Effect.Effect.Context<S["run"]>, ImmediateProvidedContexts<S, Inputs>>

interface Step<Name extends string, A, E, R, Inputs extends AnyStep[]> {
	name: Name;
	readonly inputs: Inputs;
	run: Effect.Effect<A, E, R>;
	read: Effect.Effect<A, E, StepContext<Name>>;

	[Info]?: {
		Inputs: Inputs;
		A: A;
		E: E;
		// R: R;
	};
}

type StepContext<Name extends string> = `step/${Name}`;

// biome-ignore lint/suspicious/noExplicitAny: no one cares
type AnyStep = Step<any, any, any, any, any[]>;
// biome-ignore lint/suspicious/noExplicitAny: no one cares
type Tail<Ts> = Ts extends [any, ...infer T] ? T : [];

type _ParentStepContexts<Ss extends AnyStep[], Current = never> = {
	empty: Current;
	nonempty: _ParentStepContexts<
		[...Tail<Ss>, ...Ss[0]["inputs"]],
		StepContext<Ss[0]["name"]> | Current
	>;
}[Ss extends [] ? "empty" : "nonempty"];
export type ParentStepContexts<S extends AnyStep> = _ParentStepContexts<
	S["inputs"]
>;

export class Current extends Effect.Tag("@step/__current__")<
	Current,
	{
		readonly title: string;
	}
>() {}

const getDependencies = (s: AnyStep) =>
	Effect.gen(function* () {
		const dependencies = new Map<
			AnyStep,
			{
				deps: Set<AnyStep>;
				whenResolved: Deferred.Deferred<unknown, unknown>;
			}
		>();
		const stack = [s];
		const visited = new Set<AnyStep>();

		while (true) {
			const item = stack.pop();
			if (!item) break;
			visited.add(item);

			const current = dependencies.get(item) || {
				deps: new Set<AnyStep>(),
				whenResolved: yield* Deferred.make<unknown, unknown>(),
			};
			dependencies.set(item, current);

			for (const input of item.inputs) {
				current.deps.add(input);

				if (!visited.has(input)) {
					stack.push(input);
				}
			}
		}
		return dependencies;
	});

export const run = <S extends AnyStep>(
	step: S,
): Effect.Effect<
	Effect.Effect.Success<S["run"]>,
	Effect.Effect.Error<S["run"]>,
	Exclude<Effect.Effect.Context<S["run"]>, ParentStepContexts<S>>
> =>
	// @ts-expect-error
	Effect.gen(function* () {
		yield* Effect.yieldNow();
		const dependencies = yield* getDependencies(step);
		let context = Context.empty();

		yield* Effect.forEach(
			dependencies,
			([step, { deps, whenResolved }]) =>
				Effect.gen(function* () {
					if (deps.size) {
						yield* Effect.forEach(
							deps,
							(dep) =>
								Effect.gen(function* () {
									const dependency = yield* Effect.fromNullable(
										dependencies.get(dep),
									).pipe(Effect.orDieWith(() => "Dependency not found?!"));
									yield* dependency.whenResolved;
								}).pipe(Effect.provide(context)),
							{
								discard: true,
							},
						);
					}

					yield* step.run.pipe(
						Effect.provide(context),
						Effect.tap((value) => {
							const ctx = step.read as Context.Tag<unknown, unknown>;
							context = Context.add(context, ctx, value);
						}),
						Effect.onExit((exit) => Deferred.done(whenResolved, exit)),
						Effect.annotateLogs({ step: step.name }),
					);
				}),
			{ concurrency: "unbounded" },
		);
	});

export const toDot = (step: AnyStep) =>
	Effect.gen(function* () {
		const dependencies = yield* getDependencies(step);

		const graph = digraph((g) => {
			const nodes = {} as Record<string, NodeModel>;
			for (const [step] of dependencies) {
				nodes[step.name] = g.node(step.name);
			}
			for (const [step, { deps }] of dependencies) {
				for (const dep of deps) {
					g.edge([nodes[dep.name], nodes[step.name]]);
				}
			}
		});

		return toGraphvizDot(graph);
	});
