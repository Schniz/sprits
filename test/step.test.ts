import { Effect } from "effect";
import * as Step from "../src/step";
import { test, expect, expectTypeOf } from "vitest";

test("workflow works", async () => {
	class Srv extends Effect.Service<Srv>()("Srv", {
		succeed: { value: "hi" },
	}) {}
	const order = [] as string[];
	const push = Step.Current.pipe(
		Effect.andThen(({ title }) => {
			order.push(title);
			return title;
		}),
	);
	const grandparent1 = Step.make({
		title: "grandparent1",
		inputs: [],
		run: Effect.zipLeft(push, Srv),
	});
	const parent1 = Step.make({
		title: "parent1",
		inputs: [grandparent1],
		run: push,
	});
	const grandparent2 = Step.make({
		title: "grandparent2",
		inputs: [],
		run: push,
	});
	const parent2 = Step.make({
		title: "parent2",
		inputs: [grandparent2],
		run: push,
	});
	const child = Step.make({
		title: "child",
		inputs: [parent1, parent2],
		run: Effect.zipLeft(push, Effect.all([parent2, parent1])),
	});

	const effect = Step.run(child);

	expectTypeOf(effect).toEqualTypeOf<Effect.Effect<string, never, Srv>>();

	const result = await Effect.runPromise(
		effect.pipe(Effect.provide(Srv.Default)),
	);
	expect(result).toBe("child");

	expect(order).toEqual([
		"grandparent2",
		"grandparent1",
		"parent2",
		"parent1",
		"child",
	]);

	expect(Effect.runSync(Step.toDot(child))).toMatchSnapshot();
});

test("invalid dependency tree", async () => {
	const order: string[] = [];
	const push = Step.Current.pipe(
		Effect.andThen(({ title }) => {
			order.push(title);
			return title;
		}),
	);
	// a -> b -> d
	// c --------^ but reads from a

	const a = Step.make({ title: "a", run: push, inputs: [] });
	const b = Step.make({ title: "b", run: push, inputs: [a] });
	const c = Step.make({
		title: "c",
		// @ts-expect-error
		run: Effect.andThen(a, push),
		inputs: [],
	});
	const d = Step.make({ title: "d", run: Effect.void, inputs: [b, c] });

	const effect = Step.run(d).pipe(Effect.andThen(() => order));

	const promise = Effect.runPromise(effect);
	expect(promise).rejects.toThrow("Service not found: @sprits/a");
});

test("composition", async () => {
	const order = [] as string[];
	const push = Effect.andThen(Step.Current, ({ title }) => {
		order.push(title);
		return title;
	});
	const end = Effect.andThen(Step.Current, ({ title }) => {
		order.push(`/${title}`);
		return title;
	});
	const subWorkflow = (prefix: string) => {
		const a = Step.make({ title: `${prefix}_a`, run: push, inputs: [] });
		const b = Step.make({
			title: `${prefix}_b`,
			run: Effect.andThen(a, push),
			inputs: [a],
		});
		return Step.run(b);
	};
	const first = Step.make({
		title: "first",
		run: Effect.andThen(push, subWorkflow("first")).pipe(Effect.andThen(end)),
		inputs: [],
	});
	const second = Step.make({
		title: "second",
		run: Effect.andThen(push, subWorkflow("second")).pipe(Effect.andThen(end)),
		inputs: [],
	});
	const accumulate = Step.make({
		title: "accumulate",
		run: Effect.void,
		inputs: [first, second],
	});

	await Effect.runPromise(Step.run(accumulate));
	expect(order).toEqual([
		"second",
		"first",
		"second_a",
		"first_a",
		"second_b",
		"first_b",
		"/second",
		"/first",
	]);
});
