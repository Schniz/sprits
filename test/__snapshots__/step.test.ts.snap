// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`workflow works 1`] = `
"digraph {
  "child";
  "parent2";
  "grandparent2";
  "parent1";
  "grandparent1";
  "parent1" -> "child";
  "parent2" -> "child";
  "grandparent2" -> "parent2";
  "grandparent1" -> "parent1";
}"
`;
